machine/v1/stable.controlplanemachineset.openstack.testsuite.yaml

```diff
@@ -71,6 +71,98 @@ tests:
               platform: OpenStack
               openstack: {}
     expectedError: "spec.template.machines_v1beta1_machine_openshift_io.failureDomains.openstack in body must be of type array: \"object\""
+  - name: Should accept an OpenStack failure domain with an empty failure domain array
+    initial: |
+      apiVersion: machine.openshift.io/v1
+      kind: ControlPlaneMachineSet
+      spec:
+        selector:
+          matchLabels:
+            machine.openshift.io/cluster-api-machine-role: master
+            machine.openshift.io/cluster-api-machine-type: master
+        template:
+          machineType: machines_v1beta1_machine_openshift_io
+          machines_v1beta1_machine_openshift_io:
+            metadata:
+              labels:
+                machine.openshift.io/cluster-api-machine-role: master
+                machine.openshift.io/cluster-api-machine-type: master
+                machine.openshift.io/cluster-api-cluster: cluster
+            spec:
+              providerSpec: {}
+            failureDomains:
+              platform: OpenStack
+              openstack: []
+    expected: |
+      apiVersion: machine.openshift.io/v1
+      kind: ControlPlaneMachineSet
+      spec:
+        replicas: 3
+        state: Inactive
+        strategy:
+          type: RollingUpdate
+        selector:
+          matchLabels:
+            machine.openshift.io/cluster-api-machine-role: master
+            machine.openshift.io/cluster-api-machine-type: master
+        template:
+          machineType: machines_v1beta1_machine_openshift_io
+          machines_v1beta1_machine_openshift_io:
+            metadata:
+              labels:
+                machine.openshift.io/cluster-api-machine-role: master
+                machine.openshift.io/cluster-api-machine-type: master
+                machine.openshift.io/cluster-api-cluster: cluster
+            spec:
+              providerSpec: {}
+            failureDomains:
+              platform: OpenStack
+              openstack: []
+  - name: Should accept an OpenStack spec without failureDomains
+    initial: |
+      apiVersion: machine.openshift.io/v1
+      kind: ControlPlaneMachineSet
+      spec:
+        selector:
+          matchLabels:
+            machine.openshift.io/cluster-api-machine-role: master
+            machine.openshift.io/cluster-api-machine-type: master
+        template:
+          machineType: machines_v1beta1_machine_openshift_io
+          machines_v1beta1_machine_openshift_io:
+            metadata:
+              labels:
+                machine.openshift.io/cluster-api-machine-role: master
+                machine.openshift.io/cluster-api-machine-type: master
+                machine.openshift.io/cluster-api-cluster: cluster
+            spec:
+              providerSpec: {}
+            failureDomains:
+              platform: ""
```

Comment:
What I meant was
```suggestion
```
But you can't do that in the Go serialisation, this needs to be moved to a pointer to allow that.

I guess you're trying to generate a config without any failure domains right?
I thought we had tested this already and had it working?

Reply:
>But you can't do that in the Go serialisation

exactly.

As for the tests: that's what I am trying to do but failing so far.

Reply:
>I guess you're trying to generate a config without any failure domains right?

Not exactly. What I ultimately want is my Control plane Machines to have `availabilityZone: ""`, which is the default. I don't have any preconception how to get there

Reply:
If you completely omit `failureDomains`, then you specify whatever else you want within the `providerSpec`, so you'd put `availabailityZone: ""` inside the providerSpec
In this case, CPMS doesn't do any failure domain logic and always applies the bare `providerSpec` as it was passed to it

Reply:
>If you completely omit failureDomains

Do you mean by having this?

```yaml
            failureDomains:
              platform: ""
```

As it turns out, the test passed and openshift/api is fine with it.

Reply:
Ahh yes, that works then, https://github.com/openshift/api/blob/b88ff67980ac47988df06ff9ae13d9a8e287adb7/machine/v1/0000_10_controlplanemachineset.crd.yaml#L281

In which case that's the way forward

Reply:
I knew there was a way to do it because we sent it through QE for vSphere 😅

Reply:
Not beautiful, but I'll take! Thank you.